In [2]:
import numpy as np
import pandas as pd
cohort_df= pd.read_csv ("1.1 - MLAI+C3+Cohort+data.csv")
cohort_df.head()
cohort_df.describe()

cohort_df['dfexp'] = cohort_df['YearsOfExp'].str.extract('(\d+)')
cohort_df['dfexp'] = cohort_df.dfexp.astype(int)
S = cohort_df.dfexp.sum()
avg = S / cohort_df.dfexp.count()
print(avg)

cohort_df['AvgYearsOfExp'] = cohort_df['YearsOfExp'].map(lambda x : x.rstrip('+')).astype(int)
cohort_df.head()
cohort_df.AvgYearsOfExp.mean()

7.657894736842105


7.657894736842105

In [4]:
# Expected Loss
# Now, the bank wants to find the total expected loss. This is done by calculating the expected losses 
# for each of the 10,000 customers and then adding them up.
# The data for all the customers of the bank is given in the file Inferential Statistics - Student Loan.csv. 
# What would be the value of the total expected loss?

import numpy as np
import pandas as pd
df = pd.read_csv ("3.1 - Inferential Statistics - Student Loan.csv")
# df.head()
# df.describe()

df['Recovery (%)'] = df['Recovery (%)'].str.replace('%','')
df['Recovery (%)'] = df['Recovery (%)'].astype(str).astype(float)
df['Recovery (%)'] = df['Recovery (%)'].apply(lambda x:x/100)

xloss = (df['Probability of Default'] * df['Exposure at Default (in lakh Rs.)'] * (1 - df['Recovery (%)']))
df['Expected Loss in Lakh Rs.'] = xloss

print(df.head())
print(df['Expected Loss in Lakh Rs.'].sum())

   Customer No.  Exposure at Default (in lakh Rs.)  Recovery (%)  \
0             1                              11.50        0.2000   
1             2                               0.24        0.0510   
2             3                               0.04        0.2486   
3             4                              13.81        0.0229   
4             5                              19.84        0.0347   

   Probability of Default  Unnamed: 4  Unnamed: 5  Expected Loss in Lakh Rs.  
0                  0.0070         NaN         NaN                   0.064400  
1                  0.0033         NaN         NaN                   0.000752  
2                  0.0022         NaN         NaN                   0.000066  
3                  0.0066         NaN         NaN                   0.089059  
4                  0.0020         NaN         NaN                   0.038303  
493.39131440619997


Now, let’s say that we define the random variable X as Kriti’s net winnings if she plays this game and bets £100 on the number 5. What is the expected value of X?
There are two possible values of X, 3,500 and -100. The probability of Kriti’s net winnings being 3,500 is basically the probability of her winning the game, which, as we saw in a previous question, is 1/37. The probability of her net winnings being equal to -100 is basically the probability of her losing the game, which is 36/37. So, the expected value of X = P(X=-100)*(-100) + P(X=3500)*3500 = (36/37)*(-100)+(1/37)*(3500) = -2.70.

In [5]:
print(3500*(1/37)-100*(36/37))

-2.702702702702709


The only difference between the American roulette wheel and the one you saw before, i.e. the European roulette wheel, is that, in addition to the numbers 0 to 36, it contains a number ‘00’. The player can bet on any number from 1 to 36.
Now, again, let’s define the random variable X as Kriti’s net winnings if she plays this game and bets £100 on the number 5. What is the expected value of X this time?

 Again, there are two possible values of X, 3,500 and -100. However, this time, the probability of Kriti’s net winnings being 3,500 is 1/38, as there are a total of 38 possible outcomes. The probability of her net winnings being equal to -100, basically the probability of her losing the game, is now 37/38. So, the expected value of X = P(X=-100)*(-100) + P(X=3500)*3500 = (37/38)*(-100)+(1/38)*(3500) = -5.26.

In [6]:
pxWin = 1/38   #P(X=3500)
print(pxWin)
pxLoss = 37/38 #P(X=-100) 
print(pxLoss)
# ExpectValX = P(X=-100) * (-100) + P(X=3500)*3500
ExpectValX = pxWin*3500 - pxLoss*100
print(ExpectValX)

0.02631578947368421
0.9736842105263158
-5.26315789473685


# binomial probability distribution, uniform probability distribution, Cumulative probability


binomial distribution is applicable in situations where there are a fixed number of yes or no questions, with the probability of a yes or a no remaining the same in all the questions.
 
 e.g. 
 Now, the test is modified by adding 5 true and false questions to the set of 10 MCQs. The student does not know the answer to the new questions either and randomly selects true or false for each of them too.

Now, according to binomial distribution, the probability that she would still get 5 of the 15 questions (5 true/false Qs, and 10 MCQs with 4 options) correct is:

Binomial probability cannot be used here
 For 10 of the questions (MCQs), the probability of success is p = 1 / 4. However, for the remaining 5 questions (T/F questions), the probability of success is p = 1 / 2. Hence, you cannot use binomial probability here, as the probability of success is not the same in all the trials.

In [7]:
pr = 3/5
print(pr)
pb = 2/5
print(pb)
rr = (3/5) * (3/5)
print("P(2 red balls) = ", rr)
bb = (2/5) * (2/5)
print("P(2 blue balls) = ", bb)
rb = (3/5) * (2/5)
print("P(1 red ball and 1 blue ball) = ", rb)
br = (2/5) * (3/5)
print("P(1 blue ball and 1 red ball) = ", br)

0.6
0.4
P(2 red balls) =  0.36
P(2 blue balls) =  0.16000000000000003
P(1 red ball and 1 blue ball) =  0.24
P(1 blue ball and 1 red ball) =  0.24


the cumulative probability of X, denoted by F(x), is defined as the probability of the variable being less than or equal to x.
In mathematical terms, you would write cumulative probability F(x) = P(X<x). For example, F(4) = P(X<4), F(3) = P(X<3).

You can Use Excel or R for the calculations for this part. Use "COMBIN(n,r)" in Excel and "choose(n,r)" in R to find nCr

All data that is normally distributed follows the 1-2-3 rule. This rule states that there is a -
1. 68% probability of the variable lying within 1 standard deviation of the mean
2. 95% probability of the variable lying within 2 standard deviations of the mean
3. 99.7% probability of the variable lying within 3 standard deviations of the mean

In [8]:
onesigma = 68
twosigma = 95
threesigma = 99.7
ths = (threesigma-onesigma)/2
res = onesigma+ths
print(str(round(res/100, 2)))

0.84


In [9]:
z1 = (25.2 - 35) / 5
z2 = (44.8 - 35) / 5
print(z1, z2)
ans = 0.975 - 0.025 
print(ans*100,"%")

-1.9600000000000002 1.9599999999999995
95.0 %


Not only that, you can also use Excel or Python to find the cumulative probability for Z. For example, let’s say you want to find the cumulative probability for Z = 1.5. In Excel, you would type:

 

= NORM.S.DIST(1.5, TRUE)

Basically, the syntax is:

= NORM.S.DIST(z, TRUE)

 

Here, z is the value of the Z score for which you want to find the cumulative probability. TRUE = find cumulative probability, FALSE = find probability density.

 

Also, you can find the probability without standardising. Let’s say that X is normally distributed, with mean (μ) = 35 and standard deviation (σ) = 5. Now, if you want to find the cumulative probability for X = 30, you would type:

 

= NORM.DIST(30, 35, 5, TRUE)

Basically, the syntax is:

= NORM.DIST(x, mean, standard_dev, TRUE)

What is the probability of a normally distributed random variable lying within 1.65 standard deviations of the mean?
[You can use the Z table here.]

You have to find the probability of the variable lying between μ-1.65σ and μ+1.65σ. i.e. P(μ-1.65σ < X < μ+1.65σ). In terms of Z, this becomes P(-1.65 < Z < +1.65). 
This would be equal to P(1.65) - P(-1.65) = 0.95 - 0.05 = 0.90.

What is the standard deviation of our sampling distribution? (The data for the 100 sample means is given in the CSV file upgrad_samples.csv).

0.4248

 Using the mean from the previous question, you can find the standard deviation using the formula given at https://www.mathsisfun.com/data/standard-deviation-formulas.html
. Note that you would divide by n and not n-1 as you have the data for all 100 entries of the distribution and you don't need to sample the distribution. Now, using the formula, you get that the value of the standard deviation of the distribution is 0.4248.

In [10]:
import numpy as np
import pandas as pd
df = pd.read_csv ("5.1 - Inferntial Statistics - UpGrad Samples.csv")

SampleMean = np.array(df['Sample Mean'])
print("Mean = ", SampleMean.mean())
print("Standerd deviation = ", SampleMean.std(ddof=1)) #for sample standard deviation, supply an optional ddof parameter to std()

Mean =  2.3480000000000003
Standerd deviation =  0.42698851497340296


So, the central limit theorem says that, for any kind of data, provided a high number of samples has been taken, the following properties hold true:

Sampling distribution’s mean (μX) = Population mean (μ)

Sampling distribution’s standard deviation (Standard error) = σ√n

For n > 30, the sampling distribution becomes a normal distribution

In [11]:
mean = ((1*42)+(0*58))/100
#So, the standard deviation is equal to its square root, i.e. 0.496 or 49.6%
sd = np.sqrt(((((1 - 0.42)**2)*42) + (((0 - 0.42)**2)*58))/99)
print("mean =", mean)
print("sd = ", sd)

mean = 0.42
sd =  0.4960449637488582


In [12]:
x = (((1.65*.2)/np.sqrt(10000))*100)
print("90% confidence interval is between", 50.5-x, "-and-", 50.5+x)
x = (((1.96*.2)/np.sqrt(10000))*100)
print("95% confidence interval is between", 50.5-x, "-and-", 50.5+x)
x = (((2.58*.2)/np.sqrt(10000))*100)
print("99% confidence interval is between", 50.5-x, "-and-", 50.5+x)

90% confidence interval is between 50.17 -and- 50.83
95% confidence interval is between 50.108 -and- 50.892
99% confidence interval is between 49.984 -and- 51.016


In [13]:
MOE = ((1.96*100)/np.sqrt(100))
print(MOE)
print("95% confidence interval is between", 530-MOE, "-and-", 530+MOE)
MOE = ((1.65*100)/np.sqrt(100))
print(MOE)
print("90% confidence interval is between", 530-MOE, "-and-", 530+MOE)

19.6
95% confidence interval is between 510.4 -and- 549.6
16.5
90% confidence interval is between 513.5 -and- 546.5


In [14]:
MOE = ((1.96*49.6)/np.sqrt(100))
print(MOE)
print("95% confidence interval for BJP is between", 58-MOE, " and ", 58+MOE)
MOE = ((1.96*49.6)/np.sqrt(100))
print(MOE)
print("95% confidence interval for INC is between", 42-MOE, " and ", 42+MOE)

9.721599999999999
95% confidence interval for BJP is between 48.278400000000005  and  67.7216
9.721599999999999
95% confidence interval for INC is between 32.278400000000005  and  51.721599999999995


Not make any claims regarding the election results for ward 75N
The percentage of voters that voted for BJP, falls within the interval (48.28%, 67.72%). The interval for the percentage of voters that voted for INC is (32.28%, 51.72%). Hence, anything is possible. It is possible that BJP gets 67.72% of the votes and wins the election, but it is also possible that INC could get 51.72% of the votes and win the election for ward 75N. Hence, you cannot decide the winner based on this information.

In [15]:
# As a data analyst working for Ghar ki Baat, you are tasked with finding the average rent for all 
# 2 BHK fully furnished flats in Powai, Mumbai, using this data.

import numpy as np
import pandas as pd
df = pd.read_csv ("7.1 Inferential Statistics - Powai Flats Rent.csv")
df.describe()
# What is the 90% confidence interval for the monthly rent in rupees?

rentalData = np.array(df['Monthly Rent'])
mean = rentalData.mean()
sd = rentalData.std(ddof=1) # shold be 7457, so for the sample standard deviation, supply an optional ddof parameter to std()
N = df['Monthly Rent'].count()
print("Mean = ", mean)
print("Standard deviation = ", sd)
print("N = ", N)

MOE = ((1.65*sd)/np.sqrt(N)) #margin of error
print("MOE = ", MOE)
print("90% confidence interval for rental is between", round(mean-MOE), " and ", round(mean+MOE))

MOE = ((1.96*sd)/np.sqrt(N)) #margin of error
print("MOE = ", MOE)
print("95% confidence interval for rental is between", round(mean-MOE), " and ", round(mean+MOE))

MOE = ((2.58*sd)/np.sqrt(N)) #margin of error
print("MOE = ", MOE)
print("99% confidence interval for rental is between", round(mean-MOE), " and ", round(mean+MOE))


Mean =  45571.0
Standard deviation =  7457.521851178448
N =  200
MOE =  870.0886048494972
90% confidence interval for rental is between 44701.0  and  46441.0
MOE =  1033.5597972757666
95% confidence interval for rental is between 44537.0  and  46605.0
MOE =  1360.5021821283049
99% confidence interval for rental is between 44210.0  and  46932.0
